SLING-12636: Resource Resolver: add test coverage for background init

reschke opened the pull request.

joerghoh commented on src/test/java/org/apache/sling/resourceresolver/impl/mapping/VanityPathMapEntriesTest.java:
```diff
@@ -149,6 +154,36 @@ public void setup() throws Exception {
+    private void initializeVanityPaths() throws IOException {
+        getVanityPathsProcessed().set(false);
+        mapEntries.initializeVanityPaths();
+        waitForBgInit();
```
do I get it right, that you added the ``waitForBgInit()`` just in preparation for the next change, where you will run tests also with ``isVanityPathCacheInitInBackground = true``?

reschke replied:
The tests (now) run with bginit as well, see parameters of the test class.

reschke commented on src/test/java/org/apache/sling/resourceresolver/impl/mapping/VanityPathMapEntriesTest.java:
```diff
@@ -149,6 +154,36 @@ public void setup() throws Exception {
+
+    // get vanity paths (after waiting for bg init to complete)
+    private void initializeVanityPaths() throws IOException {
+        getVanityPathsProcessed().set(false);
```
to check whether background init has finished; otherwise the expections about the number of entries in the map will be incorrect.

rombert replied:
I still don't follow the full logic; you are replacing `doInit` with changing an internal flag; this IMO makes the test assume too much about the behaviour of `MapEntries` and is going to make evolving this class more difficult.

Have you considered adding a new protected/package-private method that will encapsulate the re-reading of vanity paths?

reschke replied:
I'll do that in later PRs.

This change is focused on improving test coverage without refactoring MapEntries. Tests will be updated when refactoring is done.

FWIW, doinit() actually initializes aliases, not vanity paths. All of this will be improved step by step.

reschke replied:
> Why do you need to reset the internal state here?

I'll check.

reschke replied:
Ok. The reason is that the internal state of `MapEntries` did not account for `loadVanityPaths` called twice.

I added that, and was able to remove `getVanityPathsProcessed().set(false);` from the test.

Your point about extending `MapEntries` is understood; however the point of *this* change was to improve test coverage before doing any changes of what's being tested.

reschke commented:
@rombert , @joerghoh - I plan to go ahead with this once https://issues.apache.org/jira/browse/SLING-12637 (release of 1.12.4) is done.

reschke commented:
Need to re-open as the test fails when not run in isolation.

reschke commented:
Turns out that the actual test *causing* the failure is still unknown. That said, we have fixed the failing test to properly setup ManEntries, adjusted the test expections and added explanations. See https://github.com/apache/sling-org-apache-sling-resourceresolver/commit/c983862efb39420a6ca4e0f9b958141e5aecadbf
